reconcile_test.go

```diff
@@ -330,6 +352,226 @@ func TestReconcile(t *testing.T) {
+				Servers: map[raft.ServerID]*ServerState{
+					"96be11f3-c9b9-45ab-a719-dc9472ada6fe": {
+						Server: Server{
+							ID:          "96be11f3-c9b9-45ab-a719-dc9472ada6fe",
+							Name:        "node1",
+							Address:     "198.18.0.1:8300",
+							NodeStatus:  NodeAlive,
+							Version:     "1.9.0",
+							RaftVersion: 3,
+						},
+						State:  RaftLeader,
+						Health: ServerHealth{Healthy: true},
+					},
+					"4b92b892-ee0d-4644-84fb-3117448a0401": {
+						Server: Server{
+							ID:          "4b92b892-ee0d-4644-84fb-3117448a0401",
+							Name:        "node2",
+							Address:     "198.18.0.2:8300",
+							NodeStatus:  NodeAlive,
+							Version:     "1.9.0",
+							RaftVersion: 3,
+						},
+						State:  RaftVoter,
+						Health: ServerHealth{Healthy: true},
+					},
+					"0a79bbf7-7113-4947-a257-6179326f188c": {
+						Server: Server{
+							ID:          "0a79bbf7-7113-4947-a257-6179326f188c",
+							Name:        "node3",
+							Address:     "198.18.0.3:8300",
+							NodeStatus:  NodeAlive,
+							Version:     "1.9.0",
+							RaftVersion: 3,
+						},
+						State:  RaftVoter,
+						Health: ServerHealth{Healthy: false},
+					},
+					"b8508007-68d5-42c9-92a6-28686676867e": {
+						Server: Server{
+							ID:          "b8508007-68d5-42c9-92a6-28686676867e",
+							Name:        "node4",
+							Address:     "198.18.0.4:8300",
+							NodeStatus:  NodeAlive,
+							Version:     "1.9.0",
+							RaftVersion: 3,
+						},
+						State:  RaftVoter,
+						Health: ServerHealth{Healthy: false},
+					},
+					"1f07052e-53c7-4f99-9cb6-6bb5b39ce8a8": {
+						Server: Server{
+							ID:          "1f07052e-53c7-4f99-9cb6-6bb5b39ce8a8",
+							Name:        "node5",
+							Address:     "198.18.0.5:8300",
+							NodeStatus:  NodeAlive,
+							Version:     "1.9.0",
+							RaftVersion: 3,
+						},
+						State:  RaftNonVoter,
+						Health: ServerHealth{Healthy: true},
+					},
+					"bcd603a7-18e2-48c6-ac60-167e1556f4b0": {
+						Server: Server{
+							ID:          "bcd603a7-18e2-48c6-ac60-167e1556f4b0",
+							Name:        "node6",
+							Address:     "198.18.0.6:8300",
+							NodeStatus:  NodeAlive,
+							Version:     "1.9.0",
+							RaftVersion: 3,
+						},
+						State:  RaftNonVoter,
+						Health: ServerHealth{Healthy: true},
+					},
+				},
+				Voters: []raft.ServerID{
+					"96be11f3-c9b9-45ab-a719-dc9472ada6fe",
+					"4b92b892-ee0d-4644-84fb-3117448a0401",
+					"0a79bbf7-7113-4947-a257-6179326f188c",
+					"b8508007-68d5-42c9-92a6-28686676867e",
+				},
+			},
+			changes: RaftChanges{
+				Promotions: []raft.ServerID{
+					// both promotions should happen
+					"1f07052e-53c7-4f99-9cb6-6bb5b39ce8a8",
+					"bcd603a7-18e2-48c6-ac60-167e1556f4b0",
+				},
+				Demotions: []raft.ServerID{
+					// no demotions should happen in this reconciliation
+					// since we are starting with an even number of voters
+					"0a79bbf7-7113-4947-a257-6179326f188c",
+					"b8508007-68d5-42c9-92a6-28686676867e",
```

Comment:
The changes in this test are an input to the reconcile function, not the expected outcome we assert against. The assertions are the mock functions being called. (I am not sure if you knew that or not, but this test wasn't the easiest to parse for me, so I am clarifying this just in case 😄.)

I now see that this test case is confusing, since that set of changes would possibly never be passed to `reconcile()` in reality, but I added these to assure that the function logic ignores any demotions when we have an even number of votes.

I think the overall test will be a bit easier to reason about once I implement the changes you suggested in the comment above (simpler UUIDs and helper functions for creating state). I will also add more comments to make the intent clearer.